internal/resource/url.go

```diff
@@ -332,6 +341,68 @@ func (f *Fetcher) fetchFromDataURL(u url.URL, dest io.Writer, opts FetchOptions)
 	return f.decompressCopyHashAndVerify(dest, bytes.NewBuffer(url.Data), opts)
 }
 
+// FetchFromAWSMetadata fetches metadata from the `IMDSv2` service if its configured, else
+// it will fallback to the `IMDSv1`. By default, the value of url is an IMDSv1 endpoint
+// i.e. http://169.254.169.254/2019-10-01/user-data.
+func (f *Fetcher) FetchFromAWSMetadata(u url.URL, opts FetchOptions) ([]byte, error) {
+	userdataLatestURL := url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "latest",
+	}
+	token, err := f.fetchAWSIMDSV2Token(userdataLatestURL.String())
+	if err != nil && err != ErrIMDSV2 {
+		return nil, err
+	}
+	if err != nil && err == ErrIMDSV2 {
+		// Do nothing
+		f.Logger.Info("IMDSv2 service is disabled; fallback to IMDSv1")
+	}
+	if len(token) > 0 && err == nil {
+		opts = FetchOptions{
+			Headers: http.Header{
+				"X-aws-ec2-metadata-token": []string{token},
+			},
+		}
+		u.Path = path.Join(userdataLatestURL.Path, "user-data")
+	}
+	return f.FetchToBuffer(u, opts)
+}
+
+// fetchAWSIMDSV2Token fetches a session token from an EC2 instance (if configured
+// to use `IMDSv2`), otherwise, it will return an error.
+func (f *Fetcher) fetchAWSIMDSV2Token(u string) (string, error) {
+	if f.client == nil {
+		logger := log.New(true)
+		f.Logger = &logger
+		if err := f.newHttpClient(); err != nil {
+			return "", err
+		}
+	}
+	// check for f.Offline and return `ErrNeedNet` early to avoid
+	// a network-related issue.
+	if f.Offline {
+		return "", ErrNeedNet
+	}
+	url, _ := url.Parse(fmt.Sprintf("%s/api/token", u))
+	opts := FetchOptions{
+		Headers: http.Header{
+			"x-aws-ec2-metadata-token-ttl-seconds": []string{"21600"},
+		},
+		HTTPVerb: "PUT",
+	}
+	token, err := f.FetchToBuffer(*url, opts)
+	if err != nil {
+		if err == ErrNotFound {
+			f.Logger.Debug("cannot read IMDSv2 session token from %q", url.String())
+			return "", ErrIMDSV2
+		}
+		f.Logger.Debug("unexpected error retrieving IMDSv2 token: %v", err)
```

Comment:
```suggestion
		f.Logger.Debug("unexpected error retrieving IMDSv2 session token: %v", err)
```
For consistency with the `ErrNotFound` message.